### PR TITLE
chore(metadata): reworks labels and annotations to be their own types

### PR DESCRIPTION
**Pull request by bartoszmajsak**
## Summary

Previous implementation of commonly used metadata such as labels and annotations was based on [anonymous structs](https://github.com/maistra/odh-platform/blob/675d77bb9d46a6f9fcadce804e8e34dae3d2e2a6/pkg/metadata/annotation.go) which only contained a key for such an entry. It was a convenient way of grouping keys and exposing them as easily discoverable names (i.e. `metadata.Labels.OwnerName`) but it required additional funcs to make them applicable as `metadata.Options` when modifying cluster resources.

This proposed solution is inspired by controller-runtime client's Option structs. Both labels and annotations are type aliases of `string` enabling simple syntax while constructing them, e.g. `owner := labels.OwnerName("dashboard")`, while at the same time implementing `metadata.Option` interface (previously `func` alias type) making it immediately available to apply as part cluster operations:

```golang
if errApply := cluster.Apply(ctx, r.Client, resources, labels.OwnerName("dashboard"), labels.OwnerKind("Dashboard")); errApply != nil {
		return fmt.Errorf("could not apply routing resources for type %s: %w", exportMode, errApply)
}
```

Additionally, when needed to be part of the map to pass further (for example as `LabelSelector`) these new types provide methods to access their keys and values like so:

```golang
routingExportedLabel := labels.RoutingExported("true")
ownerNameLabel := labels.OwnerName(ownerName)
ownerKindLabel := labels.OwnerKind(ownerKind)

listOpts := []client.ListOption{
	client.InNamespace(target.GetNamespace()),
	client.MatchingLabels{
		routingExportedLabel.Key(): routingExportedLabel.Value(),
		ownerNameLabel.Key():       ownerNameLabel.Value(),
		ownerKindLabel.Key():       ownerKindLabel.Value(),
	},
}
```

In order to get only the key for a given label/annotation type, the following construct can be used:

```golang
ownerNameLabelKey := labels.OwnerName("").Key()
```

this effectively creates a label instance of
`platform.opendatahub.io/owner-name: ""`.

## Additional changes

For better discoverability through code-completion they have been promoted to their own packages `pkg/metadata/{annotations,labels}`.

## Further improvements

Instead of copy-paste approach when creating a new label, we could leverage code generation.

That would reduce:

```golang
// OwnerName is the name of the owner of the resource.
type OwnerName string

func (o OwnerName) ApplyToMeta(obj metav1.Object) {
	addLabel(o, obj)
}

func (o OwnerName) Key() string {
	return "platform.opendatahub.io/owner-name"
}

func (o OwnerName) Value() string {
	return string(o)
}
```

to

```golang
// OwnerName is the name of the owner of the resource.
// +opendatahub:label:key=platform.opendatahub.io/owner-name
type OwnerName string
```

and at the same time ensure consistent implementation.

### CODE REVIEW

**Comment by bartoszmajsak** [comment]
Small weekend side-quest :)

File: pkg/metadata/annotations/types.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+func (a AuthorizationGroup) Key() string {
+	return "security.opendatahub.io/authorization-group"
+}
+
+func (a AuthorizationGroup) Value() string {
+	return string(a)
+}
+
+// RoutingExportMode defines the export mode for the routing capability. It can be
+// either "internal" or "external" or both, delimited by ";".
```
Comment (cam-garrison) [inline pkg/metadata/annotations/types.go]:
```suggestion
// either "public" or "external" or both, delimited by ";".
```

File: controllers/routing/routing_reconcile_resources.go [hunk 2]
```diff
@@ -8,14 +8,16 @@ import (
...
+	"github.com/opendatahub-io/odh-platform/pkg/metadata/labels"
 	"github.com/opendatahub-io/odh-platform/pkg/spi"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	k8stypes "k8s.io/apimachinery/pkg/types"
 )
 
 func (r *PlatformRoutingReconciler) reconcileResources(ctx context.Context, target *unstructured.Unstructured) error {
-	// TODO shouldn't we make it a predicate for ctrl watch instead?
+	// TODO shouldn't we make it a predicate for ctrl watch instead?x
```
Comment (cam-garrison) [inline controllers/routing/routing_reconcile_resources.go]:
```suggestion
	// TODO shouldn't we make it a predicate for ctrl watch instead?
```

**Review by cam-garrison — CHANGES_REQUESTED** [review]
some nits on comments added but looks good!

File: controllers/routing/routing_delete_resources.go [hunk 3]
```diff
@@ -33,14 +34,14 @@ func (r *PlatformRoutingController) HandleResourceDeletion(ctx context.Context,
...
 
 func (r *PlatformRoutingController) deleteOwnedResources(ctx context.Context, target *unstructured.Unstructured, gvkList []schema.GroupVersionKind) error {
-	ownerName := target.GetName()
-	ownerKind := target.GetObjectKind().GroupVersionKind().Kind
-	ownerUID := string(target.GetUID())
+	ownerName := labels.OwnerName(target.GetName())
+	ownerKind := labels.OwnerKind(target.GetObjectKind().GroupVersionKind().Kind)
+	ownerUID := labels.OwnerUID(target.GetUID())
 
 	resourceOwnerLabels := client.MatchingLabels{
```
Comment (aslakknutsen) [inline controllers/routing/routing_delete_resources.go]: Could we get a helper alla: labels.Match(labels...) client.MatchingLabels here somewhere?

Constructing the label type and "manually" constructing the matcher based on the key and value that "hopefully" match (easy copy paste error) isn't too far of the way it already worked.

Comment (bartoszmajsak) [reply controllers/routing/routing_delete_resources.go]: Yeah I was thinking about it too today when looking at your latest PR. Will do

Comment (bartoszmajsak) [reply controllers/routing/routing_delete_resources.go]: So now it's

```golang
labels.MatchingLabels(
		labels.OwnerName(target.GetName()),
		labels.OwnerKind(target.GetObjectKind().GroupVersionKind().Kind),
		labels.OwnerUID(target.GetUID()),
)
````

Comment (bartoszmajsak) [reply controllers/routing/routing_delete_resources.go]: Thanks for the idea

**Review by cam-garrison — APPROVED** [review]
gave this another look, lgtm :+1:

**Review by aslakknutsen — APPROVED** [review]

**Pull request merged**